contrib/storage-jdbc/src/main/java/org/apache/drill/exec/store/jdbc/JdbcCatalogSchema.java

```diff
@@ -53,11 +54,12 @@
     try (Connection con = source.getConnection();
          ResultSet set = con.getMetaData().getCatalogs()) {
       connectionSchemaName = con.getSchema();
-      while (set.next()) {
-        final String catalogName = set.getString(1);
-        CapitalizingJdbcSchema schema = new CapitalizingJdbcSchema(
-            getSchemaPath(), catalogName, source, dialect, convention, catalogName, null, caseSensitive);
-        schemaMap.put(schema.getName(), schema);
+      if (!ClickhouseConstant.PRODUCT_NAME.equals(con.getMetaData().getDatabaseProductName())) {
```

Comment:
Let's think about this approach a bit. There is no comment here (there should be), so I'll infer that Clickhouse somehow does something special. Cool. So I'm working on Postgres, or DB X and I want to do something similar. Do we end up with a bit set of if-statements? If I work on DB X, how do I ensure I don't break Clickhouse, since I can't run that code?

Can we think about a way of abstracting this kind of thing into a DB-specific helper class? The default helper does "stock" JDBC. Clickhouse (and my DB X) code go into DB-specific helpers. Now, to get DB X to work, I don't have to worry about breaking Clickhouse.

See if you can work out a "mini-plugin" for such code. Happy to provide suggestions if you like.

Reply:
I have added JdbcSchemaFactory and JdbcSqlGenerator, please have a look.